src/middleware/authUser.js

```diff
@@ -0,0 +1,36 @@
+const boom = require('boom')
+
+const roles = require('../constants/roles.js')
+const { auth } = require('../constants/errors.json')
+
+const User = require('../models/User')
+
+// function to check role of user matches minRole allowed on the route
+const verifyRole = ({ minRole }) => user => {
+  const userScope = user.scope
+
+  if (minRole === roles.BASIC) return
+  if (minRole === roles.ADMIN && userScope !== roles.BASIC) return
+  if (minRole === roles.SUPER && userScope === roles.SUPER) return
+
+  return Promise.reject(boom.unauthorized(auth.UNAUTHORIZED))
+}
+
+// Check user from the token exists, and add their id to the req.user for potential use in permissioning
+const addUserIdToSession = req => {
+  return User.find({ username: req.user.username }).then(users => {
+    if (users.length === 0) {
+      return Promise.reject(boom.unauthorized(auth.UNAUTHORIZED))
+    }
+
+    req.user.id = users[0].id
+    return req.user
+  })
+}
+
+module.exports = opts => (req, res, next) => {
+  addUserIdToSession(req)
```

Comment:
Err kind of although a couple of things:

> 1. We can add the username to the jwt.

Yeah, we already do - I'm not entirely sure why you said this? I was asking why we don't add the id to the JWT at the same time, instead of adding it later during the 'auth' stage

Also, with regards to your flow, what is authUser.js doing? it's both attaching user objects to the request (the auth part) and then verifying the role (i.e. doing some sort of permissioning). But then the authSession.js also does part of your auth (authenticates token, attaches what's in it to the request)

The specific flow we implemented:
1. Make a login POST request, response contains token in cookie, which contains the username and user role
2. Attempt to access secure route with token in cookie
3. authSession.js authenticates token, and puts token into req.user 
4. authUser.js checks user exists, and adds user ID to request
5. authUser.js verifies role and allows or disallows

Is authUser in this instance just an example of the auth/permission layers in your flow?